## **qPCR analysis web-interface**
The qPCR analysis web-interface is designed to conveniently and intuitively plot amplification plots and melting curves. In addition, it automates the calculation of the relative gene expression, calculates the efficiency of every primerset used,  and plots this data in a clear way. Both the LightCycler480 and BioRad systems are suitable, but require different preparation of the Excel files

## **Analysis of the LightCycler480 output**
Before using the qPCR-analysis website, the data coming out of the LightCycler480 needs to be transposed. Use the http://humgen.nl/scripts/transpose/ website and use the manual written before by Maurice Overzier.
After copying the data to excel, remove the data from empty wells (wells of the plate where you did not pipet anything in). Name this sheet ‘Data’. Name every sample as following: cell_primer (eg. Control1_GAPDH).

<img src="https://github.com/bjbvoesenek/qPCR-analysis/blob/update-manual/Markdown images/LinReg_Data.png">      

Go to LinRegPCR and use the ‘Determine Baselines’ function, according to the manual of Maurice Overzier. This will create ‘Data_output’ and ‘Data_compact’. Do not change the names of these sheets or any of the column names within these sheets.

<img src="https://github.com/bjbvoesenek/qPCR-analysis/blob/update-manual/Markdown images/LinReg_Data_compact.png"> 

If you exported melting curves from the LightCycler480, copy the content of the .txt file to your Excel file in a new sheet named ‘Melting curves’. Delete columns (both X and Sample) of empty wells.

<img src="https://github.com/bjbvoesenek/qPCR-analysis/blob/update-manual/Markdown images/LinReg_Melting curves.png"> 

Go to https://humgen.nl/scripts/qPCR-analysis/ and upload your Excel(*.xlsx) file that you want to analyze. Press ‘Submit’ and follow further instructions on the website.

<img src="https://github.com/bjbvoesenek/qPCR-analysis/blob/update-manual/Markdown images/qPCR_web_interface.png"> 

Your results will be ready and downloaded within a minute. Your download will consist of three folders: ‘Input’, ‘Figures’, and ‘Data’. The Input folder contains your input file, the housekeeping genes, and normalizer cell lines selected for this analysis. The ‘Figures’ folder contains figures generated by the script. You will find amplification plots (qPCR_plots_sorted.PDF) and the melting curves (Melting_curves_sorted.PDF). Both these figures are sorted in a way that every row is a cell line, and every column a gene. This makes it easy to compare cell lines and see which samples are different.

<img src="https://github.com/bjbvoesenek/qPCR-analysis/blob/update-manual/Markdown images/Amplification curves.png"> 

Other figures include bargraphs showing the triplicate average Ct value of each sample, the relateive expression values of each sample, and the efficiency of the primersets used. Control lines are represented by gray bars, other lines by blue bars.
The ‘Data’ folder contains Excel sheets with the average Ct per condition or the relative gene expression per condition, so that you can make figures yourself as well. 



## **Analysis of the BioRad output**
Before uploading the excel file to the website, there are a few steps you need to do. Firstly, rename the sheet containing the Cq values to ‘Cq’. In the column ‘Sample’, name the samples containing the mq control ‘mq’ or ‘MQ’. If you exported melting curves, copy the content to a new sheet in your excel file that you are going to analyze. Rename this sheet to ‘Melting curves’.
 
<img src="https://github.com/bjbvoesenek/qPCR-analysis/blob/update-manual/Markdown images/BioRad_Cq.png">

<img src="https://github.com/bjbvoesenek/qPCR-analysis/blob/update-manual/Markdown images/BioRad_Melting curves.png"> 

Go to https://humgen.nl/scripts/qPCR-analysis/ and upload your Excel(*.xlsx) file that you want to analyze. Press ‘Submit’ and follow further instructions on the website.

<img src="https://github.com/bjbvoesenek/qPCR-analysis/blob/update-manual/Markdown images/qPCR_web_interface.png"> 

Your results will be ready and downloaded within a minute. Your download will consist of three folders: ‘Input’, ‘Figures’, and ‘Data’. The Input folder contains your input file, the housekeeping genes, and normalizer cell lines selected for this analysis. The ‘Figures’ folder contains figures generated by the script. You will the melting curves (Melting_curves_sorted.PDF) here. This figure sorted in a way that every row is a cell line, and every column a gene. This makes it easy to compare cell lines and see which samples are different.

<img src="https://github.com/bjbvoesenek/qPCR-analysis/blob/update-manual/Markdown images/Amplification curves.png"> 

Other figures include bargraphs showing the triplicate average Ct value of each sample and the relateive expression values of each sample. Control lines are represented by gray bars, other lines by blue bars.
The ‘Data’ folder contains Excel sheets with the average Ct per condition or the relative gene expression per condition, so that you can make figures yourself as well. 

## Installation

qPCR-analysis is developed in Python 3.6.13 and requires the following python libraries:
- matplotlib.pyplot 3.3.4
- natsort 8.2.0
- xlrd 1.2.0
- pandas 0.23.0


In addition, it requires Python's build in libraries:
- argparse
- math
- os
- statistics
- sys

### Web interface

qPCR analysis can also be operated from a web browser.
A PHP-based web interface is included in the `web-wrapper` directory.
To enable this interface, install a webserver such as Apache, enable the PHP
 module, and add this project to a location where the webserver can reach it.
Make sure the `web-wrapper/data` directory is writable by the webserver user.

If you are not using an Apache webserver, make sure you protect the
data directory from direct downloads, to ensure the safety of the data.
For Apache users, a `.htaccess` file is already in place to block access.

#### Maintenance of the web interface

For every analysis, a new directory is created
 in the `web-wrapper/data` directory.
To avoid filling up your drive, make sure old data folders are removed.
You can use the cronjob below for this purpose.
Make sure the cron will run as the user of the webserver,
 and that you replace `path` with the appropriate directory.

```
0 0 * * * find /path/data/ -type d -iname 1\* -mtime +0 | xargs rm -r
```

This removes data folders older than 24 hours, every night at midnight.
Feel free to change the settings; this cron can easily be run at any given time,
 run more than once a day, or less frequently, e.g., once a week or month.
If you want the cronjob to create output, add a "v" at the end of the command,
 i.e., `xargs rm -rv`.



## How to use qPCR-analysis


### Through the terminal

You can invoke qPCR-analysis by running:

```
python3 qpcr_analysis --input <input file> \
                      --genes <gene> [<gene> [...]] \
                      --controls <control> [<control> [...]] 
```

This will create three directories in the current directory;
 `Input`, `Figures`, and `Data`.
qPCR should take a little less than a minute to process your data.
Results will be created in all three directories.

***TODO: Explain each file?***

If you'd like to have a list of primers and samples listed in your input file,
 you can invoke qPCR-analysis by running:
```
python3 qpcr_analysis --input <input file>
```

When invoked without any further arguments, qPCR-analysis will read out your
 spreadsheet and generate `Input/Genes.txt` and `Input/Cell_lines.txt` in the
 current directory.
These files, respectively, contain the list of primers and samples mentioned in
 your given spreadsheet.
Values from these lists can then be used for the `--genes` and `--controls`
 arguments.


### Web interface

The web interface is designed to work as intuitively as possible.
An example installation is currently available at
 [HumGen.nl/scripts/qPCR-analysis](https://humgen.nl/scripts/qPCR-analysis/).
The web interface does not require any log in, but simply asks you
 to upload the `.xlsx` input file that was obtained from LinReqPCR.
After pressing the submit button, the interface runs the qPCR-analysis script to
 extract the list of primers and samples, and then lets you choose the
 housekeeping genes and the controls cell lines.
In each case, the interface shows buttons for each possible value.
Clicking the buttons you'd like to choose, selects them.
For each, you must select at least one value.
Then, press the submit button to continue.
After selecting both at least one housekeeping gene and then at least one
 control sample, the script will take a while to execute - please wait for it
 to finish execution.
This will take a little less than a minute.
When the qPCR-analysis script is done, the webinterface will create a ZIP
 archive containing all the files (the input file, the settings, and all output
 files) and send it to the web browser for download.
If any failures occur, clear error messages will indicate at what point the
 script failed to run, so that actions can be taken to fix these.
